k-distribution/include/builtin/domains.k

```diff
@@ -444,39 +444,40 @@ module INT-SYMBOLIC [symbolic, kore]
   imports  INT-COMMON
   imports ML-SYNTAX
 
-  /**
-   * Definability conditions
-   */
-
-  rule
-    #Ceil(@I1:Int /Int @I2:Int)
-  =>
-    {(@I2 =/=Int 0) #Equals true} #And #Ceil(@I1) #And #Ceil(@I2)
-  [anywhere]
-
-  rule
-    #Ceil(@I1:Int %Int @I2:Int)
-  =>
-    {(@I2 =/=Int 0) #Equals true} #And #Ceil(@I1) #And #Ceil(@I2)
-  [anywhere]
-
-  rule
-    #Ceil(@I1:Int modInt @I2:Int)
-  =>
-    {(@I2 =/=Int 0) #Equals true} #And #Ceil(@I1) #And #Ceil(@I2)
-  [anywhere]
-
-  rule
-    #Ceil(@I1:Int >>Int @I2:Int)
-  =>
-    {(@I2 >=Int 0) #Equals true} #And #Ceil(@I1) #And #Ceil(@I2)
-  [anywhere]
-
-  rule
-    #Ceil(@I1:Int <<Int @I2:Int)
-  =>
-    {(@I2 >=Int 0) #Equals true} #And #Ceil(@I1) #And #Ceil(@I2)
-  [anywhere]
+  // Definability Conditions
+  rule #Ceil(@I1:Int /Int   @I2:Int) => {(@I2 =/=Int 0) #Equals true} #And #Ceil(@I1) #And #Ceil(@I2) [anywhere]
+  rule #Ceil(@I1:Int %Int   @I2:Int) => {(@I2 =/=Int 0) #Equals true} #And #Ceil(@I1) #And #Ceil(@I2) [anywhere]
+  rule #Ceil(@I1:Int modInt @I2:Int) => {(@I2 =/=Int 0) #Equals true} #And #Ceil(@I1) #And #Ceil(@I2) [anywhere]
+  rule #Ceil(@I1:Int >>Int  @I2:Int) => {(@I2 >=Int 0)  #Equals true} #And #Ceil(@I1) #And #Ceil(@I2) [anywhere]
+  rule #Ceil(@I1:Int <<Int  @I2:Int) => {(@I2 >=Int 0)  #Equals true} #And #Ceil(@I1) #And #Ceil(@I2) [anywhere]
+
+  // Arithmetic Normalization
+  rule I +Int 0 => I [simplification]
+  rule I -Int 0 => I [simplification]
+
+  rule I +Int B => B +Int I          [concrete(I), symbolic(B), priority(51), simplification]
+  rule A -Int I => A +Int (0 -Int I) [concrete(I), symbolic(A), priority(51), simplification]
```

Comment:
What is this `priority` attribute intended to do?

Reply:
It's to make sure the rules preceding it are applied first:

```
  rule I +Int 0 => I [simplification]
  rule I -Int 0 => I [simplification]

  rule I +Int B => B +Int I          [concrete(I), symbolic(B), priority(51), simplification]
  rule A -Int I => A +Int (0 -Int I) [concrete(I), symbolic(A), priority(51), simplification]
```

For example, if you have `X -Int 0`, then you want it to go directly to `X` (and not to `X +Int (0 -Int 0)`).